database/sqlserver/sqlserver.go

```diff
@@ -344,3 +374,16 @@ func (ss *SQLServer) ensureVersionTable() (err error) {
 
 	return nil
 }
+
+func getMSITokenProvider(resource string) (func() (string, error), error) {
```

Comment:
Do you mind briefly walking me through the decision to use a higher order function here? What is it buying us?

Reply:
Ohh, I see in [this PR](https://github.com/denisenkom/go-mssqldb/pull/546/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R125) it's due to the short-lived nature of AAD tokens and this function ensures a refresh.

Reply:
mssql.NewAccessTokenConnector expects a callback which is used to [build the connection](https://github.com/denisenkom/go-mssqldb/blob/35162394aa7974ac37c4b42ddb1c200f6b01ebac/accesstokenconnector.go#L14) and that will be used to retrieve fresh tokens. [msi.EnsureFresh()](https://github.com/Azure/go-autorest/blob/10e0b31633f168ce1a329dcbdd0ab9842e533fb5/autorest/adal/token.go#L882) will update the token if it is about to expire, and then msi.OAuthToken() gets the token to be used for the connection